src/library/scala/collection/mutable/ArrayBuffer.scala

```diff
@@ -232,6 +239,25 @@ final class ArrayBufferView[A](val array: Array[AnyRef], val length: Int) extend
 /** An object used internally by collections backed by an extensible Array[AnyRef] */
 object RefArrayUtils {
 
+  // We chose 8 as a good size for the minimum array
+  private[this] val minimalLength: Long = 8L
+
+  def reduceSize(array: Array[AnyRef], n: Int): Array[AnyRef] = {
+    if (n < 0)
+      throw new Exception(s"Collections can not have a negative size")
+    if (n == Int.MaxValue)
+      throw new Exception(s"Collections can not have more than ${Int.MaxValue} elements")
```

Comment:
`n == MaxValue` shouldn't be a problem here but `arrayLength == MaxValue` needs to be taken into account to ensure we shrink down to a power of 2.